pkg/kubelet/nodestatus/setters.go

```diff
@@ -197,36 +198,65 @@ func NodeAddress(nodeIP net.IP, // typically Kubelet.nodeIP
 			// For steps 3 and 4, IPv4 addresses are preferred to IPv6 addresses
 			// unless nodeIP is "::", in which case it is reversed.
 			if nodeIPSpecified {
-				ipAddr = nodeIP
+				ipAddr = []net.IP{nodeIP}
 			} else if addr := net.ParseIP(hostname); addr != nil {
-				ipAddr = addr
+				ipAddr = []net.IP{addr}
 			} else {
 				var addrs []net.IP
 				addrs, _ = net.LookupIP(node.Name)
 				for _, addr := range addrs {
 					if err = validateNodeIPFunc(addr); err == nil {
-						if isPreferredIPFamily(addr) {
-							ipAddr = addr
-							break
-						} else if ipAddr == nil {
-							ipAddr = addr
+						if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+							ipAddr = append(ipAddr, addr)
+						} else {
+							if isPreferredIPFamily(addr) {
+								ipAddr = []net.IP{addr}
+								break
+							} else if len(ipAddr) < 1 {
+								ipAddr = []net.IP{addr}
+							}
```

Comment:
I would say that the better behavior would be that regardless of feature gate, we should use all of the returned IPs, but sort the `isPreferredIPFamily` ones first. (There is code that does that in the cloud provider case above which could possibly be shared with a little bit of rewriting.)

Reply:
I would generally agree, but that _would_ be a behaviour change.

Reply:
yeah, it's a possibility, my point about the breaking change is that we can't return a different IP address and I don't know how the rest of components are consuming these addresses